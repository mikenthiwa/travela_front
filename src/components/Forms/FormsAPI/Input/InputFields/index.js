import HtmlInput from './HtmlInput';
import DropdownSelect from './DropdownSelect';
import DateInput from './Date';
import filterDropdownSelect from './filterDropdownSelect';
import ButtonToggler from './ButtonToggler';
import NumberInput from './NumberInput';
import CheckBox from './CheckBox';
import TextArea from './TextArea';
import TagsInput from './TagsInput';
import MultipleChoiceDropdown from './MultipleChoiceDropdown';
import SelectChoiceDropdown from './SelectChoiceDropdown';
import OnboardingButtonToggler from './OnboardingButtonTogggle';

export {
  DropdownSelect,
  DateInput,
  ButtonToggler,
  HtmlInput,
  NumberInput,
  filterDropdownSelect,
  CheckBox,
  TextArea,
  TagsInput,
  MultipleChoiceDropdown,
  SelectChoiceDropdown,
  OnboardingButtonToggler
};
