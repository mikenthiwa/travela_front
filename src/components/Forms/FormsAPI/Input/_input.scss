form .form-input {
  display: flex;
  flex-flow: column;
  height: 64px;
  margin: 0 0 20px 0;
  transition: height 0.5s ease-out;
    @media (max-width: 450px) {
    width: 100%;
    height: 90px;
    .arrival-date {
      width: calc(100% - 0px) !important;
      .calendar-icon {
        right: 12px !important;
      }
    }
  }
  @media (max-width: 348px) {
    font-size: 12px;
  }
  &.full-width {
    width: 100%;
    height: 64px;
    transition: height 0.4s ease-out;
    &.other-dest {
      &--hidden,
      &--hiddenerror {
        height: 0px;
        overflow: hidden;
        margin: 0;
      }
      &--error {
        height: 84px;
      }
    }
  }
  span.error {
    color: red;
    font-size: 12px;
  }
  &.error {
    height: 84px;
    .input,
    input, textarea {
      border: 1px solid #ff5359;
    }
    .input:hover,
    input:hover, textarea:hover {
      border: 1px solid #ff5359;
    }
    .input:focus,
    input:focus, textarea:focus {
      box-shadow: 0px 0px 3.5px rgba(255, 83, 89, 0.7);
      border: 1px solid #ff5359;
    }
  }
  input,
  .input, textarea {
    box-sizing: border-box;
    width: 100%;
    height: 36px;
    border-radius: 2px;
    outline: none;
    font-size: 14px;
    line-height: 18px;
    color: #4f4f4f;
    margin-top: 8px;
    font-family: "DIN Pro";

    cursor: pointer;
    &[type="text"] {
      cursor: text;
    }
    button {
      width: 50%;
      margin: 0;
      &.bg-onboarding {
        &--inactive {
          background-color: #ffffff;
          color: #4f4f4f;
        }
        &--active {
          border-radius: 0 2px 2px 0;
          background-color: #E2E5EB;
          color: #3359DB
        }
      }
    }
  }
  input:hover, textarea:hover,
  .input.select:hover {
    border: 1px solid rgba(51, 89, 219, 0.7);
  }
  input:focus, textarea:focus,
  .input.select:focus {
    box-shadow: 0px 0px 3.5px rgba(51, 89, 219, 0.6);
    border: 1px solid rgba(51, 89, 219, 0.7);
  }
  input, textarea,
  .input.select {
    border: 1px solid #d4d4d4;
    box-sizing: border-box;
    background: white;
    padding-left: 15px;
    position: relative;

    &::placeholder {
      color: #999999;
      @media(max-width: 450px) {
        font-size: 10.5px;
      }
    }
    .value {
      position: relative;
      display: flex;
      align-items: center;
      height: 100%;
      .select-dropdown {
        position: absolute;
        right: 15px;
        transition: transform 0.05s ease-out;
        transform-origin: center;
        @media(max-width: 390px) {
          margin-right: -11px;
        }
      }
      @media(max-width: 340px) {
        font-size: 11px;
      }
      @media(max-width: 450px) {
        width: 100% !important;
      }
    }
  }
  .input.select.manager-select {
    width: 260px !important;
  }

  .novisibility{
    visibility: hidden;
}

  label {
    span {
      color: #999999;
      font-family: "DIN Pro";
      font-size: 12px;
      line-height: 15px;
    }
  }
  .select-menu {
    z-index: 150;
    position: absolute;
    left: -1.05px;
    top: 26px;
    padding-left: 0px;
    margin-top: 0;
    box-sizing: border-box;
    display: flex;
    flex-flow: column;
    width: calc(100% + 3px);
    background-color: #ffffff;
    list-style-type: none;
    box-shadow: 0px 0px 1px 1px rgba(0, 0, 0, 0.1);
    border-radius: 2px;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    transition: transform 0.4s ease-out;
    transform-origin: top;
    box-sizing: border-box;
    max-height: 110px;
    overflow-y: scroll;
    overflow-x: hidden;
    li {
      align-items: center;
      cursor: pointer;
      height: 32px;
      &:hover {
        background-color: #f8f8f8;
        color: #3359db;
      }
      div {
        display: flex;
        align-items: center;
        padding-left: 15px;
        outline: none;
        width: 100%;
      }
    }
  }
  .date-wrapper, .date-wrapper > div {
    width: 100%;
    display: flex;
    position: relative;
    input {
      background: white;
      color: #999999;
      width: 190px;
      cursor: text;
      @media (max-width: 450px) {
        width: 100%;
      }
    }

    .input.select .select-menu {
        transition-delay: 0.25s;
        /* Don't reduce this */
    }

    .input.select .select-menu {
        transition-delay: 0.25s;
        /* Don't reduce this */
    }

    .input.select .select-menu {
        transition-delay: 0.25s;
        /* Don't reduce this */
    }
    .calendar-icon {
      position: absolute;
      right: 15px;
      top: 20px;
    }
    .react-datepicker-wrapper, .react-datepicker__input-container {
        width: 100%;
    }
  }
  input:disabled {
    cursor: not-allowed;
    &:hover {
      border: 1px solid #d4d4d4;
    }
  }
  .input.select .select-menu {
    transition-delay: 0.25s;
    /* Don't reduce this */
  }
  .input.select .select-menu {
    transform: scaleY(0);
  }
  .input.select:focus .select-menu {
    transform: scaleY(1);
  }
  .react-datepicker-popper {
    box-shadow: 0.5px 0px 1px 0px rgba(0, 0, 0, 0.1);
    margin: 0;
    position: absolute;
    z-index: 10;
  }
  .react-datepicker-popper[data-placement^="top"] {
    margin: 0;
    top: 5px !important;
  }
  .react-datepicker-popper[data-placement^="bottom"] {
    margin: 0;
    bottom: 1px !important;
  }
  .calendar-body {
    font-family: "DIN Pro";
    border: 1px solid #e4e4e4;
    border-radius: 2px;
    background-color: #ffffff;
    .calendar-day {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 36.5px;
      width: 35.5px;
      border-right: 1px solid #e4e4e4;
      border-radius: 0;
      box-sizing: border-box;
      background-color: #ffffff;
      margin: 0;
      font-size: 14px;
      color: #4f4f4f;
      &:first-child {
        border-left: 1px solid #e4e4e4;
      }
    }
    .react-datepicker__header {
      background-color: #ffffff;
      color: #3359db;
      border: none;
    }
    .react-datepicker__month {
      margin: 0;
      padding: 5px;
    }
    .react-datepicker__triangle {
      display: none;
    }
    .react-datepicker__week {
      box-sizing: border-box;
      display: flex;
      width: 250px;
      border-top: 1px solid #e4e4e4;
      &:last-child {
        border-bottom: 1px solid #e4e4e4;
      }
    }
    .react-datepicker__day-names {
      display: flex;
      padding-left: 5px;
      .react-datepicker__day-name {
        display: flex;
        justify-content: center;
        width: 36.5px;
        height: 16px;
        margin: 0;
        color: #4f4f4f;
        font-family: "DIN Pro";
        font-size: 12px;

        line-height: 15px;
      }
    }
    .react-datepicker__day--disabled {
      cursor: default;
      color: #ccc !important;
    }
    .react-datepicker__current-month {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 64px;
      background-color: #ffffff;
      color: #3359db;
    }
    .react-datepicker__navigation {
      top: 32px;
      outline: none;
      &--next {
        border-left-color: #3359db;
      }
      &--previous {
        border-right-color: #3359db;
      }
    }
    .react-datepicker__day {
      box-sizing: border-box;
      &--selected {
        background-color: #f8f8f8;
        color: #3359db;
      }
      &--outside-month {
        background-color: #ffffff;
        color: #e4e4e4;
      }
    }
  }
}

.required-field {
  margin-left: 61%;
  font-size: 9px;
  color: red;
  font-family: DIN Pro;
  @media (max-width: 479px) {
    margin-left: 40%;
  }
  @media (max-width: 320px) {
    margin-left: 34%;
  }
}

input[type=number]
{
  -moz-appearance: textfield;
}

input[type="number"]::-webkit-outer-spin-button,
input[type="number"]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  background: #fff url() no-repeat center center;
  width: 1.5em;
  height: 50px;
  border-left: 0px solid white;
  opacity: 0.5; /* shows Spin Buttons per default (Chrome >= 39) */
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
}

.style-number {
  display: grid;
  float: right;
  margin: -28px 8px 41px 10px;
  margin-bottom: -33px;
}

.image1 {
  transform: rotate(181deg);
  margin: 0px 0px 5px 0px;
  pointer-events: none;
}

.image2 {
  transform: rotate(358deg);
  pointer-events: none;
}

/* The checkbox container */
.container {
  display: block !important;
  position: relative;
  padding-left: 35px;
  margin-bottom: 12px;
  cursor: pointer;
  font-size: 14px !important;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

/* Hide the browser's default checkbox */
.container input {
  position: absolute;
  display: none;
  cursor: pointer;
  height: 0;
  width: 0;
}

/* Create a custom checkbox */
.checkmark {
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: #eee;
}

/* On mouse-over, add a grey background color */
.container:hover input ~ .checkmark {
  background-color: #ccc;
}

/* When the checkbox is checked, add a blue background */
.container input:checked ~ .checkmark {
  background-color: #3359DB;
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark:after {
  content: "";
  position: absolute;
  display: none;
}

/* Show the checkmark when checked */
.container input:checked ~ .checkmark:after {
  display: block;
}

/* Style the checkmark/indicator */
.container .checkmark:after {
  left: 9px;
  top: 5px;
  width: 5px;
  height: 10px;
  border: solid white;
  border-width: 0 3px 3px 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.textarea-box {
  resize: none;
  outline: none;
  overflow: auto;
  min-height: 60px;
}

.error{
  border: red;
}

.react-datepicker {

  &__year{
    &-dropdown {
      background: white;
      padding: 1em;
      box-shadow: 1px 1px 1px 1px rgba(0, 0, 0, 0.1);
    }

    &-option {
      font-size: 1.3em;

      &:first-child:hover{
        background: none;
        transform: translateY(-1px);
      }
      &:last-child:hover{
        background: none;
        transform: translateY(1px);
      }

      &:hover {
        background: rgba(0, 0, 0, .1);
      }
    }
  }

  &__navigation{
    &--years {
      top: 0 !important;
      border: none;
      width: 20px;
      height: 20px;
      background: #3359DB;
      position: relative;
      clip-path: polygon(50% 0%, 100% 50%, 50% 28%, 0% 50%);

      &-previous {
        transform: rotate(180deg);
      }
    }
  }
}

.form-input{
  &__textarea {
    position: relative;
    &__character-counter {
      position: absolute;
      bottom: -15px;
      text-align: right;
      font-size: 10px;
      right: 3px;
      color: rgba(0,0,0,.6)
    }

    &--error {
      .form-input__textarea__character-counter {
        display: none;
      }
    }
  }
}
