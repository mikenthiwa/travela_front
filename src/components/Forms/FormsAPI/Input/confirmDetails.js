import React, {Component} from 'react';
import Input from './index';

class InputRenderer{
  constructor(formMetadata) {
    this.formMetadata = formMetadata;
  }

  // eslint-disable-next-line react/display-name
  createInput = props => <Input {...props} />;

  switchDetailsProps(name, type, confirmInputDetails) {
    const formMetadata = this.formMetadata || {};
    switch (type) {
    case 'filter-dropdown-select':
      confirmInputDetails.choices = this.formMetadata.dropdownSelectOptions[name];
      return confirmInputDetails;
    case 'onboarding-button-toggler':
      confirmInputDetails.choices = (formMetadata.buttonToggleOptions || {})[name];
      return confirmInputDetails;
    default:
      return confirmInputDetails;
    }
  }
  renderInput = (name, type, customProps) => {
    const selection = customProps && customProps.selection;
    const labelKey = name.split('-')[0];
    const fieldName = (labelKey.match('arrivalDate') && (selection && !selection.match('return')))
      ? 'leavingDate' : labelKey;
    let confirmInputDetails = {
      name,
      type,
      label: this.formMetadata.confirmPersonalInputLabels[fieldName].label,
      labelNote: customProps ? customProps.labelNote :
        this.formMetadata.confirmPersonalInputLabels[fieldName].note,
      autoComplete: 'off'
    };
    customProps ? confirmInputDetails['data-parentid'] = customProps.parentid : null;
    confirmInputDetails = this.switchDetailsProps(name, type, confirmInputDetails);
    confirmInputDetails = {
      ...confirmInputDetails,
      ...customProps
    };
    return this.createInput(confirmInputDetails);
  }
}

export {Input};
export default InputRenderer;
