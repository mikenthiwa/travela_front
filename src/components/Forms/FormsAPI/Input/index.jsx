import React, { PureComponent, Fragment } from 'react';
import { PropTypes } from 'prop-types';
import {
  HtmlInput,
  DropdownSelect,
  DateInput,
  ButtonToggler,
  NumberInput,
  filterDropdownSelect,
  CheckBox,
  TextArea,
  TagsInput,
  MultipleChoiceDropdown,
  SelectChoiceDropdown,
  OnboardingButtonToggler,
} from './InputFields';
import createEventHandlersFor from '../formEventHandlers';
import './_input.scss';
import './InputFields/Date.scss';
import { RealFormContext } from '../FormContext/FormContext';

class Input extends PureComponent {

  newProps = {};

  getInputType() {
    let { name, type } = this.props;
    const { targetForm, validatorName } = this;
    // get event handlers for the form in this context, for its input named 'name'
    const eventHandlers = createEventHandlersFor(targetForm, name, validatorName);
    return this.switchTypeWithProps(type, eventHandlers);
  }

    switchTypeWithProps = (type, eventHandlers) => {
      const { onChange, minDate, maxDate } = this.props;
      switch (type) {
      case 'button-toggler':
        this.newProps = {
          ...this.props,
          onChange: onChange ||  eventHandlers.handleSelectTogglerOpt
        };
        return ButtonToggler;
      case 'onboarding-button-toggler':
        this.newProps = {
          ...this.props,
          onChange: onChange ||  eventHandlers.handleSelectTogglerOpt
        };
        return OnboardingButtonToggler;
      case 'date':
        this.newProps  = {
          ...this.props,
          minimumDate: minDate,
          maximumDate: maxDate,
          onChange: onChange || eventHandlers.handleSelectDate,
          onBlur: eventHandlers.handleInputBlur
        };
        return DateInput;
      case 'dropdown-select':
      case 'filter-dropdown-select':
        this.newProps= {
          ...this.props,
          onChange: onChange || eventHandlers.handleSelectDropdown,
        };
        return this.switchDropdownInputTypes(type);
      case 'checkbox':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleCheckBoxChange
        };
        return CheckBox;
      case 'tags':
        this.newProps = {
          ...this.props,
          handleDelete: onChange || eventHandlers.handleTagDeleted,
          handleAddition: onChange || eventHandlers.handleTagAdded,
          handleInputBlur: eventHandlers.handleTagAdded
        };
        return TagsInput;
      case 'multiple-choice-dropdown':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleInputChange,
          onBlur: eventHandlers.handleInputBlur
        };
        return MultipleChoiceDropdown;
      case 'select-choice-dropdown':
        this.newProps = {
          ...this.props,
          onChange: onChange || eventHandlers.handleInputChange,
          onBlur: eventHandlers.handleInputBlur
        };
        return SelectChoiceDropdown;
      default:
        this.newProps = {
          ...this.props,
          onBlur: eventHandlers.handleInputBlur,
          onChange: onChange ||  eventHandlers.handleInputChange
        };
        return this.switchTextBasedInputTypes(type);
      }
    }

    switchTextBasedInputTypes(type) {
      switch (type) {
      case 'text':
      case 'password':
      case 'email':
        return HtmlInput;
      case 'number':
        return NumberInput;
      case 'textarea':
        return TextArea;
      }
    }

    switchDropdownInputTypes(type) {
      switch (type) {
      case 'filter-dropdown-select':
        return filterDropdownSelect;
      default:
        return DropdownSelect;
      }
    }

    labelNote(note) { // a small message tied to an inputs label
      return note ? (
        <span>
        &nbsp;
          { note }
        </span>
      ) : null;
    }


    renderAsterisk(){
      /* This API assumes that all inputs are required and therefore a red asterisk is displayed
         *  but this is not enforced with html
         * To remedy the former, if one has a field that should be optional,
         * they have to explicitly pass 'required: false' into customProps
         * This method checks for this and will display the asterisk appropriately.
         */

      const { required, labelNote } = this.props;
      let requiredInput = required === undefined ? true : required;
      return (
        <Fragment>
          {requiredInput && ( <span style={{color: 'red'}}> * </span> ) }
          {this.labelNote(labelNote)}
        </Fragment>
      );
    }

    render() {
      return (
        <RealFormContext.Consumer>
          {({ errors, values, targetForm, validatorName }) => {
            this.targetForm = targetForm;
            this.validatorName = validatorName;

            let { name, label, className } = this.props;
            const value = values ? values[name]: '';
            const error = errors ? errors[name]: '';
            let customClass = className ? className : '';
            // switch input types into InputElement
            const InputElement = this.getInputType();
            return (
              <div
                className={`form-input ${customClass} ${error ? 'error' : ''}`}>
                <label htmlFor={name}>
                  {label}
                  {this.renderAsterisk()}
                </label>
                <InputElement value={value} error={error} {...this.newProps} />
                <span className="error">
                  {error}
                </span>
              </div>
            );
          }}
        </RealFormContext.Consumer>
      );
    }
}

Input.contextTypes = {
  errors: PropTypes.object,
  values: PropTypes.object,
  targetForm: PropTypes.object,
  validatorName: PropTypes.string
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.oneOfType([
    PropTypes.func, PropTypes.object
  ]),
  labelNote: PropTypes.string,
  className: PropTypes.string,
};

Input.defaultProps = {
  labelNote: '',
  className: '',
  onChange: null
};

export default Input;
