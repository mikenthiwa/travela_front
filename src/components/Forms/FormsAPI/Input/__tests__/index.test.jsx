import React from 'react';
import Input from '..';
import context from '../../FormContext/FormContext'


jest.mock('../../FormContext/FormContext', () => {
  return {
    RealFormContext: {
      Consumer: ({ children }) => children({
        values: { name: '' },
        validatorName: 'validator',
        errors: { name: 'This field is required'},
        targetForm: {
          validate: jest.fn(),
        }
      })
    }
  };
});

describe('<Input />', () => {
  let wrapper;

  beforeEach(() => {

    wrapper = mount(
      <Input
        name="name"
        type="text"
        onChange={()=>{}}
        label="Test label"
      />
    );
  });


  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Displays the error messge', () => {
    const errorText = wrapper.find('span.error').text();
    expect(errorText).toBe('This field is required');
  });

});
