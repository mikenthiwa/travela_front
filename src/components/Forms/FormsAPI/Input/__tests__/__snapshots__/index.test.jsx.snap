// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders as expected 1`] = `
<div
  className="form-input  error"
>
  <label
    htmlFor="name"
  >
    Test label
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       * 
    </span>
  </label>
  <input
    className="error "
    label="Test label"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value=""
  />
  <span
    className="error"
  >
    This field is required
  </span>
</div>
`;
